### PR TITLE
mgr/autoscaler: Introduce noautoscale flag

### PR DESCRIPTION
**Pull request by kamoltat**
`noautoscale` flag is a feature where the
user can choose to flip the switch between
turning autoscale `on` and `off` for all
pools with a single command.

`osd pool set noautoscale` will turn all
autoscale mode`off` for all pools.

`osd pool unset noautoscale` will turn all
autoscale mode `on` for all pools.

Address: https://tracker.ceph.com/issues/51213

Signed-off-by: Kamoltat <ksirivad@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/pybind/mgr/pg_autoscaler/module.py [hunk 1]
```diff
@@ -286,6 +293,51 @@ def set_profile_scale_down(self) -> Tuple[int, str, str]:
             self.set_module_option("autoscale_profile", "scale-down")
             return 0, "", "autoscale-profile is now scale-down"
 
+    def complete_all_progress_events(self):
```
Comment (neha-ojha) [inline src/pybind/mgr/pg_autoscaler/module.py:296]: https://jenkins.ceph.com/job/ceph-pull-requests/84714/consoleFull#1585608204c19247c4-fcb7-4c61-9a5d-7e2b9731c678

```
pg_autoscaler/__init__.py:6: note: In module imported here:
pg_autoscaler/module.py: note: In member "complete_all_progress_events" of class "PgAutoscaler":
pg_autoscaler/module.py:296: error: Function is missing a return type annotation
pg_autoscaler/module.py:296: note: Use "-> None" if function does not return a value
pg_autoscaler/module.py: note: In member "set_autoscale_mode_all_pools" of class "PgAutoscaler":
pg_autoscaler/module.py:302: error: Function is missing a type annotation
pg_autoscaler/module.py: note: In member "unset_noautoscale" of class "PgAutoscaler":
pg_autoscaler/module.py:314: error: Function is missing a return type annotation
pg_autoscaler/module.py: note: In member "set_noautoscale" of class "PgAutoscaler":
pg_autoscaler/module.py:327: error: Function is missing a return type annotation
Found 4 errors in 1 file (checked 31 source files)
```

**Comment by kamoltat** [comment]
jenkins test api

**Review by neha-ojha — COMMENTED** [review]
@kamoltat I noticed that if we run `ceph osd pool set noautoscale` and then create a new pool, the new pool has `AUTOSCALE` on. I don't think this is the behavior we want. Can you please add a test case to cover this and other scenarios?

**Comment by kamoltat** [comment]
> @kamoltat I noticed that if we run `ceph osd pool set noautoscale` and then create a new pool, the new pool has `AUTOSCALE` on. I don't think this is the behavior we want. Can you please add a test case to cover this and other scenarios?

@neha-ojha Agreed, will make the changes so that it also newly created pools will have `AUTOSCALE` off. Will create test case that covers this scenarios as well.

**Comment by kamoltat** [comment]
https://pulpito.ceph.com/ksirivad-2021-11-09_19:46:36-rados:singleton:all:pg-autoscaler.yaml-wip-ksirivad-autoscale-global-flag-distro-basic-smithi/

**Comment by kamoltat** [comment]
```
2021-11-09T20:08:42.170 INFO:tasks.workunit.client.0.smithi149.stderr:+ ceph osd pool set noautoscale
2021-11-09T20:08:42.537 INFO:tasks.workunit.client.0.smithi149.stderr:noautoscale is already set!
2021-11-09T20:08:42.546 INFO:tasks.workunit.client.0.smithi149.stderr:+ ceph osd pool create a
2021-11-09T20:08:43.544 INFO:tasks.workunit.client.0.smithi149.stderr:pool 'a' already exists
2021-11-09T20:08:43.557 INFO:tasks.workunit.client.0.smithi149.stderr:+ sleep 2
2021-11-09T20:08:45.561 INFO:tasks.workunit.client.0.smithi149.stderr:++ ceph osd pool autoscale-status
2021-11-09T20:08:45.562 INFO:tasks.workunit.client.0.smithi149.stderr:++ grep -oe off
2021-11-09T20:08:45.562 INFO:tasks.workunit.client.0.smithi149.stderr:++ wc -l
2021-11-09T20:08:45.950 DEBUG:teuthology.orchestra.run:got remote process result: 1
2021-11-09T20:08:45.951 INFO:tasks.workunit.client.0.smithi149.stderr:+ RESULT1=1
2021-11-09T20:08:45.951 INFO:tasks.workunit.client.0.smithi149.stderr:+ test 1 -eq 2
2021-11-09T20:08:45.952 INFO:tasks.workunit:Stopping ['mon/test_noautoscale_flag.sh'] on client.0...
2021-11-09T20:08:45.952 DEBUG:teuthology.orchestra.run.smithi149:> sudo rm -rf -- /home/ubuntu/cephtest/workunits.list.client.0 /home/ubuntu/cephtest/clone.client.0
2021-11-09T20:08:46.206 ERROR:teuthology.run_tasks:Saw exception from tasks.
Traceback (most recent call last):
```

**Comment by neha-ojha** [comment]
@kamoltat could you please put the test changes in a different commit and provide a link to a test run that is executing the new test?

**Comment by neha-ojha** [comment]
> @kamoltat could you please put the test changes in a different commit and provide a link to a test run that is executing the new test?

@kamoltat ping, let's try to get it merged before quincy dev freeze

**Review by neha-ojha — APPROVED** [review]
@kamoltat code changes look good, could you update https://docs.ceph.com/en/latest/rados/operations/placement-groups/ to reflect the new flag and make a note about this in the release notes

**Review by vumrao — APPROVED** [review]

**Comment by kamoltat** [comment]
Work unit passed 1/1 

However, I had to comment out the qa/tasks configuration where it sets `osd_pool_default_pg_autoscale_mode` to `off`. Somehow this messes up the stored value in the monitor,  even if we do `ceph config set global osd_pool_default_pg_autoscale_mode on`, it still doesn't really set the value to `on`

https://github.com/ceph/ceph/pull/43716/files#diff-dbf878757438981fb23ce86d3b84d3d4f1fa7d3ee627b5476b1a4edb8afda7c2R20

https://pulpito.ceph.com/ksirivad-2021-12-16_22:08:49-rados:singleton:all:test-noautoscale-flag.yaml-wip-ksirivad-autoscale-global-flag-distro-basic-smithi/

**Comment by kamoltat** [comment]
jenkins test make check

**Comment by kamoltat** [comment]
jenkins test ceph windows tests

**Comment by kamoltat** [comment]
Run Analysis on: http://pulpito.front.sepia.ceph.com/yuriw-2021-12-23_16:50:03-rados-wip-yuri6-testing-2021-12-22-1410-distro-default-smithi/
Here is the pad:
https://pad.ceph.com/p/yuriw-2021-12-23_16_50_03-rados-wip-yuri6-testing-

**Comment by ljflores** [comment]
Failures, unrelated:
    https://tracker.ceph.com/issues/53499
    https://tracker.ceph.com/issues/52124
    https://tracker.ceph.com/issues/52652
    https://tracker.ceph.com/issues/53422
    https://tracker.ceph.com/issues/51945
    https://tracker.ceph.com/issues/53424
    https://tracker.ceph.com/issues/53394
    https://tracker.ceph.com/issues/53766
    https://tracker.ceph.com/issues/53767

Details:
    Bug_#53499: test_dashboard_e2e.sh Failure: orchestrator/02-hosts-inventory.e2e failed. - Ceph - Mgr - Dashboard
    Bug_#52124: Invalid read of size 8 in handle_recovery_delete() - Ceph - RADOS
    Bug_#52652: ERROR: test_module_commands (tasks.mgr.test_module_selftest.TestModuleSelftest) - Ceph - Mgr
    Bug_#53422: tasks.cephfs.test_nfs.TestNFS.test_export_create_with_non_existing_fsname: AssertionError: NFS Ganesha cluster deployment failed - Ceph - Orchestrator
    Bug_#51945: qa/workunits/mon/caps.sh: Error: Expected return 13, got 0 - Ceph - RADOS
    Bug_#53424: CEPHADM_DAEMON_PLACE_FAIL in orch:cephadm/mgr-nfs-upgrade/ - Ceph - Orchestrator
    Bug_#53394: cephadm: can infer config from mon from different cluster causing file not found error - Ceph - Orchestrator
    Bug_#53766: ceph orch ls: setting cgroup config for procHooks process caused: Unit libpod-$hash.scope not found - Ceph - Orchestrator
    Bug_#53767: qa/workunits/cls/test_cls_2pc_queue.sh: killing an osd during thrashing causes timeout - Ceph - RADOS

**Pull request merged**